server/src/business-layer/services/StripeService.ts

```diff
@@ -452,6 +463,71 @@ export default class StripeService {
     }
   }
 
+  private async isLastSpotTaken(
+    bookingSlotService: BookingSlotService,
+    bookingDataService: BookingDataService,
+    stripe: Stripe,
+    bookingSlotId: string
+  ): Promise<boolean> {
+    const bookingSlot =
+      await bookingSlotService.getBookingSlotById(bookingSlotId)
+
+    const bookings = await bookingDataService.getBookingsBySlotId(bookingSlotId)
+    const bookingCount = bookings.length
+
+    const sessions = await stripe.checkout.sessions.list({
+      limit: 100,
+      expand: ["data.payment_intent"]
+    })
+
+    // Pending checkout sessions for this slot
+    const pendingSessions = sessions.data.filter(
+      (session) =>
+        session.metadata.booking_slot_id === bookingSlotId &&
+        (session.payment_intent as Stripe.PaymentIntent)?.status ===
+          "requires_payment_method"
+    ).length
+
+    const availableSlots =
+      bookingSlot.max_bookings - bookingCount - pendingSessions
+    return availableSlots <= 0
+  }
+
+  private async expireOtherCheckoutSessions(
+    stripe: Stripe,
+    date: string
+  ): Promise<void> {
+    try {
+      // Fetch all checkout sessions for the specific date
+      const sessions = await stripe.checkout.sessions.list({
+        limit: 100,
+        expand: ["data.payment_intent"]
+      })
+
+      const sessionsToExpire = sessions.data.filter(
+        (session) =>
+          session.metadata.dates.includes(date) &&
+          (session.payment_intent as Stripe.PaymentIntent)?.status ===
+            "requires_payment_method"
+      )
```

Comment:
You should look at some similar (but still quite different) logic (from `server\src\service-layer\controllers\PaymentController.ts`) 

```ts
      const openSessions = await stripeService.getRecentActiveSessions(
        CheckoutTypeValues.BOOKING,
        MINUTES_AGO,
        true
      )

// example if i want to check what booking slots it involves
const firstSessionBookingSlots = JSON.parse(openSessions[0].metadata[BOOKING_SLOTS_KEY]) as Array<string>

// if that contains the now fully booked booking slot then we expire the session.
```